segmentation.py

```diff
@@ -0,0 +1,575 @@
+        # distribution.
+        x = np.dot(omega, mu)
+        self.__prod[:m, :m] = omega
+        self.__prod[:m, m:] = x
+        self.__prod[m:, :m] = x.transpose()
+        self.__prod[m:, m:] = np.dot(np.transpose(mu), x) + eta * sigma
+        self.__weight = eta
+
+    def update(self, X, Y):
+        """Update the sufficient statistics given observed data.
+
+        The sufficient statistics represent the only parameters required to
+        describe the shape of the distribution. Due to conjugacy, the
+        sufficient statistics also form the hyper-parameters of the
+        distribution. In a Bayesian context, when the sufficient statistics are
+        referred to as 'hyper-parameters', it implies the sufficient statistics
+        reflect prior information about the distribution. After updating the
+        distribution with observed data, the sufficient statistics refer to the
+        posterior distribution.
+
+        """
+
+        # (Equation 5a, b)
+        #
+        #     | XX    XY |
+        #     | YX    YY |
+        #
+        if np.ndim(X) > 1:
+            k, m = np.shape(X)
+            x = np.dot(np.transpose(X), Y)
+
+            # Update the statistics given a block of data (in the following
+            # order: XX, XY, YX, YY)
+            self.__prod[:m, :m] += np.dot(np.transpose(X), X)
+            self.__prod[:m, m:] += x
+            self.__prod[m:, :m] += x.transpose()
+            self.__prod[m:, m:] += np.dot(np.transpose(Y), Y)
+            self.__weight += k
+
+        else:
+            m = np.size(X)
+            x = np.outer(X, Y)
+
+            # Update the statistics given a single datum.
+            self.__prod[:m, :m] += np.outer(X, X)
+            self.__prod[:m, m:] += x
+            self.__prod[m:, :m] += x.transpose()
+            self.__prod[m:, m:] += np.outer(Y, Y)
+            self.__weight += 1
+
+    def logconst(self):
+
+        m, n = self.__m, self.__n
+
+        # Note usage of the log-determinant 'trick':
+        #
+        #     log(det(A)) = 2*sum(log(diag(chol(A))))
+        #
+        d = np.diag(linalg.cholesky(self.__prod))
+        w = self.__weight
+
+        # Evaluate the log-normalization constant.
+        # (Equation 8)
+        return special.gammaln(0.5*(w - np.arange(n))).sum() - \
+               n * np.log(d[:m]).sum() - \
+               w * np.log(d[m:] / np.sqrt(w)).sum() - \
+               n * (0.5 * w) * np.log(0.5 * w)
+
+    def param(self):
+        """Return parameters of the posterior distribution."""
+
+        m = self.__m
+        s = linalg.cholesky(self.__prod).transpose()
+        w = self.__weight
+
+        # Compute the parameters of the posterior distribution.
+        return linalg.solve(s[:m, :m], s[:m, m:]), \
+               np.dot(s[:m, :m].transpose(), s[:m, :m]), \
+               np.dot(s[m:, m:].transpose(), s[m:, m:]) / w, \
+               w
+
+    def rand(self):
```

Comment:
`genRandParameters`? Note: we should be clear about the distinction between parameters (i.e. `A` and `B`) and hyper-parameters (i.e. `mu`, `omega', etc.) to avoid confusion.